profiling/src/lib.rs

```diff
@@ -0,0 +1,713 @@
+mod bindings;
+mod logging;
+mod profiling;
+mod sapi;
+
+use crate::profiling::Profiler;
+use bindings as zend;
+use bindings::{
+    datadog_php_str, sapi_getenv, DatadogPhpProfilingGlobals, ZendExtension, ZendResult,
+};
+use ddprof::exporter::{Tag, Uri};
+use libc::{c_char, c_int, c_ulong, c_void};
+use log::{debug, error, info, trace, warn, LevelFilter};
+use sapi::Sapi;
+use std::cell::{RefCell, RefMut};
+use std::ffi::CStr;
+use std::mem::MaybeUninit;
+use std::ops::{Deref, DerefMut};
+use std::str::FromStr;
+use std::sync::atomic::Ordering;
+use std::sync::Once;
+use std::time::Instant;
+
+static NAME: &[u8] = b"datadog-profiling\0";
+static VERSION: &[u8] = b"0.7.0\0";
+static mut SAPI: RefCell<Sapi> = RefCell::new(Sapi::Unknown);
+
+#[no_mangle]
+pub extern "C" fn get_module() -> &'static mut zend::ModuleEntry {
+    /* In PHP modules written in C, this just returns the address of a global,
+     * mutable variable. In Rust, you cannot initialize such a complicated
+     * global variable because of initialization order issues that have been
```

Comment:
would the `lazy_static` crate help you avoid rolling your own `unsafe { }` later in this function?

[looks like once_cell was also mentioned below, but one or the other seems useful]

Reply:
In this particular case, even if I switch to a lazy_static or a Once cell (which I've done locally), this is _still_ handing out a mutable reference to static variable, so it's just fundamentally unsafe. However, many other usages of `unsafe` have been reduced by using these crates.